### PR TITLE
[Enhancement] Support mv_rewrite_staleness param for mv rewrite

### PR DESCRIPTION
**Pull request by LiShuMing**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Problem Summary(Required) ：
- add a configuration setting to reduce the frequency of partition checks.
```
   // NOTE: The `maxMVRewriteStaleness` option helps you achieve consistently high performance
    // with controlled costs when processing large, frequently changing datasets.
    //
    // MV's Refresh Time = max(BaseTables' Refresh Time)
    // MV's Staleness    = now() - MV's Refresh Time
    // If MV's Staleness <= maxMVRewriteStaleness: returns data directly from the MV without reading the base tables.
    // otherwise:  reads data from the base tables.
    @SerializedName(value = "maxMVRewriteStaleness")
    private int maxMVRewriteStaleness = 0;

```

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 60 / 73 (82.19%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|14|17|82.35%|[852, 853, 854]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|46|56|82.14%|[2814, 2818, 2822, 2823, 2824, 2825, 2826, 2827, 2830, 2831]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 75 / 89 (84.27%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|44|54|81.48%|[2812, 2816, 2820, 2821, 2822, 2823, 2824, 2825, 2828, 2829]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|14|17|82.35%|[852, 853, 854]|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|9|10|90.00%|[361]|
|:large_blue_circle:|com/starrocks/alter/Alter.java|8|8|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/alter/Alter.java [hunk 1]
```diff
@@ -313,22 +313,22 @@ private void processModifyTableProperties(ModifyTablePropertiesClause modifyTabl
         int partitionTTL = INVALID;
         if (properties.containsKey(PropertyAnalyzer.PROPERTIES_PARTITION_TTL_NUMBER)) {
             partitionTTL = PropertyAnalyzer.analyzePartitionTimeToLive(properties);
-            properties.remove(PropertyAnalyzer.PROPERTIES_PARTITION_TTL_NUMBER);
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/alter/Alter.java:316]: why

File: fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java [hunk 2]
```diff
@@ -305,6 +307,17 @@ public void setLastRefreshTime(long lastRefreshTime) {
     // Maintenance plan for this MV
     private transient ExecPlan maintenancePlan;
 
+    // max delay duration to check which partitions to refresh, in this duration we use last
+    // `needToRefreshPartitionNamesCache` as the current `needToRefreshPartitionNames` result.
+    // NOTE: unit second.
+    private int maxMVRewriteStaleness = 0;
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java:313]: what's the exactly behavior of this variable ? Decide the refresh behavior, or the rewrite staleness ?

File: fe/fe-core/src/main/java/com/starrocks/alter/Alter.java (hunk 1 above)
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/alter/Alter.java:316]: I've checked that each `analyzePartitionTimeToLive ` method will remove the property. 

no need to remove it again.

File: fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java [hunk 3]
```diff
@@ -305,6 +307,17 @@ public void setLastRefreshTime(long lastRefreshTime) {
     // Maintenance plan for this MV
     private transient ExecPlan maintenancePlan;
 
+    // max delay duration to check which partitions to refresh, in this duration we use last
+    // `needToRefreshPartitionNamesCache` as the current `needToRefreshPartitionNames` result.
+    // NOTE: unit second.
```
Comment (liuyufei9527) [inline fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java:312]: actual is millisecond

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 95 / 112 (84.82%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/MaterializedViewOptimizer.java|2|3|66.67%|[38]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|22|27|81.48%|[346, 870, 871, 872, 1130]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|44|54|81.48%|[2816, 2820, 2824, 2825, 2826, 2827, 2828, 2829, 2832, 2833]|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|9|10|90.00%|[354]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewritePreprocessor.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/Alter.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMaterializedViewRefreshProcessor.java|1|1|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java [hunk 4]
```diff
@@ -305,7 +307,28 @@ public void setLastRefreshTime(long lastRefreshTime) {
...
-    public static class MvRewriteContext {
+    // The maximum delay duration is used to determine which partitions to refresh.
+    // During this duration, the current result of 'needToRefreshPartitionNamesCache' is used
+    // as the basis for 'needToRefreshPartitionNames'.
+    // Please note that the unit of measurement is in seconds.
+    @SerializedName(value = "maxMVRewriteStaleness")
+    private int maxMVRewriteStaleness = 0;
+
+    // The next timestamp, measured in seconds, determines when to check for
+    // partitions that need to be refreshed(unit: ms).
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java:318]: unit is ms or second?
the comment is ambiguous

File: fe/fe-core/src/main/java/com/starrocks/common/util/PropertyAnalyzer.java [hunk 5]
```diff
@@ -159,6 +159,8 @@ public class PropertyAnalyzer {
     public static final String PROPERTIES_FOREIGN_KEY_CONSTRAINT = "foreign_key_constraints";
     public static final String PROPERTIES_UNIQUE_CONSTRAINT = "unique_constraints";
 
+    public static final String PROPERTIES_MV_REWRITE_STALENESS = "mv_rewrite_staleness";
```
[the remark this answers is not in the document]
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/common/util/PropertyAnalyzer.java]: done.

File: fe/fe-core/src/main/java/com/starrocks/common/util/PropertyAnalyzer.java [hunk 6]
```diff
@@ -339,6 +341,23 @@ public static List<TableName> analyzeExcludedTriggerTables(Map<String, String> p
...
     }
 
+    public static int analyzeMVRewriteStaleness(Map<String, String> properties)
+            throws AnalysisException {
+        int maxMVRewriteStaleness = INVALID;
+        if (properties != null && properties.containsKey(PROPERTIES_MV_REWRITE_STALENESS)) {
+            try {
+                maxMVRewriteStaleness = Integer.parseInt(properties.get(PROPERTIES_MV_REWRITE_STALENESS));
+            } catch (NumberFormatException e) {
+                throw new AnalysisException("Partition Refresh Number: " + e.getMessage());
```
[the remark this answers is not in the document]
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/common/util/PropertyAnalyzer.java]: done.

File: fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java (hunk 4 above)
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java]: fix it.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 95 / 112 (84.82%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/MaterializedViewOptimizer.java|2|3|66.67%|[38]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|22|27|81.48%|[346, 870, 871, 872, 1130]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|44|54|81.48%|[2816, 2820, 2824, 2825, 2826, 2827, 2828, 2829, 2832, 2833]|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|9|10|90.00%|[354]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewritePreprocessor.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/Alter.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMaterializedViewRefreshProcessor.java|1|1|100.00%|[]|

**Review by ABingHuang — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 123 / 144 (85.42%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/MaterializedViewOptimizer.java|2|3|66.67%|[38]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|44|54|81.48%|[2816, 2820, 2824, 2825, 2826, 2827, 2828, 2829, 2832, 2833]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|41|50|82.00%|[345, 513, 535, 570, 578, 891, 892, 893, 1138]|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|9|10|90.00%|[354]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewritePreprocessor.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/Alter.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMaterializedViewRefreshProcessor.java|10|10|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 127 / 150 (84.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/MaterializedViewOptimizer.java|2|3|66.67%|[38]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|45|56|80.36%|[345, 513, 517, 542, 577, 583, 591, 904, 905, 906, 1151]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|44|54|81.48%|[2838, 2842, 2846, 2847, 2848, 2849, 2850, 2851, 2854, 2855]|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|9|10|90.00%|[354]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewritePreprocessor.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/Alter.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMaterializedViewRefreshProcessor.java|10|10|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 127 / 150 (84.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/MaterializedViewOptimizer.java|2|3|66.67%|[38]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|45|56|80.36%|[127, 339, 508, 512, 574, 580, 588, 901, 902, 903, 1148]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|44|54|81.48%|[2838, 2842, 2846, 2847, 2848, 2849, 2850, 2851, 2854, 2855]|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|9|10|90.00%|[354]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewritePreprocessor.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/Alter.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMaterializedViewRefreshProcessor.java|10|10|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 132 / 153 (86.27%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/MaterializedViewOptimizer.java|2|3|66.67%|[38]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|44|54|81.48%|[2838, 2842, 2846, 2847, 2848, 2849, 2850, 2851, 2854, 2855]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|43|52|82.69%|[335, 504, 508, 570, 576, 584, 897, 898, 899]|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|9|10|90.00%|[354]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewritePreprocessor.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/Alter.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMaterializedViewRefreshProcessor.java|17|17|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 130 / 151 (86.09%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/MaterializedViewOptimizer.java|2|3|66.67%|[38]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|44|54|81.48%|[2838, 2842, 2846, 2847, 2848, 2849, 2850, 2851, 2854, 2855]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|43|52|82.69%|[340, 509, 513, 575, 581, 589, 902, 903, 904]|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|9|10|90.00%|[354]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewritePreprocessor.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/Alter.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMaterializedViewRefreshProcessor.java|15|15|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 130 / 151 (86.09%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/MaterializedViewOptimizer.java|2|3|66.67%|[38]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|44|54|81.48%|[2840, 2844, 2848, 2849, 2850, 2851, 2852, 2853, 2856, 2857]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|43|52|82.69%|[341, 510, 514, 576, 582, 590, 914, 915, 916]|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|9|10|90.00%|[354]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewritePreprocessor.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/Alter.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMaterializedViewRefreshProcessor.java|15|15|100.00%|[]|

**Review by murphyatwork — DISMISSED** [review]

**Comment by mofeiatwork** [comment]
it deserves writing a document to explain this feature

File: fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java [hunk 7]
```diff
@@ -458,38 +484,66 @@ public static Expr getPartitionExpr(MaterializedView materializedView) {
...
     public Set<String> getUpdatedPartitionNamesOfOlapTable(OlapTable baseTable) {
-        Map<String, BasePartitionInfo> baseTableVisibleVersionMap = getRefreshScheme()
+        Map<String, BasePartitionInfo> mvBaseTableVisibleVersionMap = getRefreshScheme()
                 .getAsyncRefreshContext()
                 .getBaseTableVisibleVersionMap()
                 .computeIfAbsent(baseTable.getId(), k -> Maps.newHashMap());
         Set<String> result = Sets.newHashSet();
-        // check whether there are partitions added and have data
+
+        boolean isLessThanMVRewriteStaleness = isLessThanMVRewriteStaleness();
```
[the remark this answers is not in the document]
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java:494]: fix it.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 129 / 151 (85.43%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/MaterializedViewOptimizer.java|2|3|66.67%|[38]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|44|54|81.48%|[2864, 2868, 2872, 2873, 2874, 2875, 2876, 2877, 2880, 2881]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|43|52|82.69%|[341, 514, 518, 558, 581, 587, 915, 916, 917]|
|:large_blue_circle:|com/starrocks/alter/Alter.java|7|8|87.50%|[331]|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|9|10|90.00%|[354]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewritePreprocessor.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMaterializedViewRefreshProcessor.java|15|15|100.00%|[]|

**Review by ABingHuang — APPROVED** [review]

**Review by murphyatwork — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Pull request merged**